### ISSUE TITLE
Allow different black vs white visits / playouts for matches

### ISSUE THREAD

**Issue by Mardak** [gcp/leela-zero#1131]
As noted in https://github.com/gcp/leela-zero/issues/1113#issuecomment-377696237, we probably want to be able to scale distributed matches to test networks of different sizes more while accounting for their speed differences.

Instead of using a common `visits` or `playouts`, there should be a way to specify a different amount for each network.

https://github.com/gcp/leela-zero/blob/97c2f8137a3ea24938116bfbb2b0ff05c83903f0/autogtp/Management.cpp#L251-L254

At minimum, one additional option would be something like `white_visits` where the original `visits` is assumed to be for black. Alternatively, add both `white_visits` and `black_visits` and prefer each of those for the matching color and fall back to `visits`.

Alternatively, instead of giving exact values, the server could provide a ratio/multiplier. Not much of a difference of having the server compute the desired visits vs client though.

cc @marcocalignano @roy7

**Comment by marcocalignano** [comment]
But would you want that? Only because the bigger network is slower? That would put the bigger network to a disadvantage.

**Comment by Mardak** [comment]
As @gcp pointed out in https://github.com/gcp/leela-zero/issues/1113#issuecomment-376545627, the bigger network needs to "make up" for its slowness.

Allowing different visits for black and white allows distributed test matches to quantify the ELO difference for the same network as well as allows for "handicapping" of different sized network tests -- although unclear if we would give the bigger network the standard amount of visits, e.g., 3200, and the faster network a multiplier more, e.g., 16000 if using gcp's "5 times slower" estimate.

**Comment by marcocalignano** [comment]
I do not think that comments says anything about number of visits (or playouts). At the end the number of visits or playouts determines the MCTS search that should similar for both networks regardless how much time they take to evaluate them.

**Comment by Mardak** [comment]
@marcocalignano are you suggesting there should be a different way to account for the larger network being slower to compute -- i.e., not by varying the visits or playouts? From your comments in the other issue, it sounds like your answer would be "no" and that you're suggesting it's more "fair" / "equal" and that we **should** be using the same number of visits / playouts when comparing different sized networks:

>  Third having the same playouts (or visits) number for both networks guaranties (for the same reason of the second point) that the match is "equal", the bigger network will take more time to evaluate the moves but will be the same number of moves that the smaller network evaluates.

Unless I'm misunderstanding @gcp's comments, in the transition to 128x10 https://github.com/gcp/leela-zero/issues/965#issuecomment-369968805, he made a similar comment to "make up" for 10 / 6 = 1.66 increase in size over the 128x6 networks.

I suppose the alternative to fixing this issue is to continue what has been done of estimating the ELO difference of the expected slowdown and subtracting that from the measured ELO gain of the slower / larger network. I believe @jkiliani/@odeint (?) did some of those measurements before.

Although based on measured result of the initial 256x20 network over a then-best's sibling 128x10 network:
```
93229e15 VS 5716dc2a | 368 : 91 (80.17%)
93229e15: 9876 ELO
5716dc2a: 9633 ELO
```

Assuming the speed-adjusted results of 256x20 are accurate of indicating the larger network is **not** stronger, the ~4x slowdown is *at least* a 243 (9876 - 9633) ELO difference. That seems to be consistent with a previously measured ~200 ELO difference per doubling of playouts / halving of speed. Where actual win rates of the equivalent time matches indicate the 256x20 actually loses up to 80%, meaning its ELO should be closer to 9390 (9633 - 243).

**Comment by StanTraykov** [comment]
While results from #1113 (the 93229e issue + the real-world test) indicate that a network of this size might be so slow as to be weaker, despite a formidable fixed-visit winrate, is it certain that the issue should necessarily dominate architecture upgrade decisions? The results in the referenced issue also show how vast the difference is with hardware from 5 years ago, when 256x20 would have been beyond hopeless on consumer cards. If we stopped training for 5 years for some reason and wanted to continue then, the 256x20 net in this test would almost certainly *already* be the much stronger one (on future hardware) and this question would be entirely moot (at least with these exact block/filter figures). It's probably less time until the curves intersect and LZ's distributed effort itself is not a short-term exercise, so maybe it's best to stick with more hardware-independent measurements, such as fixed-visits? In the meantime, stalled improvement over N number of games with fixed visit count is a good indication that an architecture upgrade is needed--maybe even if it is not absolutely stronger in real terms? At least experience so far shows that it will quickly improve.

**Comment by marcocalignano** [comment]
We never take time into account, not in selfplay, and not in matches. That is why every time we have a bigger network gcp tell us don't worry selfplay will be slower. 
If you want to take into account the time the you have to change the way you do matches and then you have to adjust the time so that the slower (bigger) network manages to do 3200 visit (like now is set for the matches). Of course this vary for each GPU or CPU so you cannot do it unless you know which GPU you use, you cannot distribute it. I do not even think it is a sound way to test the progress of the network since the selfplay are not time constraint. The only fair thing to do is let each network all the time thay want but limit the MCTS search to the same number for both. Then the better network wins.

**Comment by marcocalignano** [comment]
@StanTraykov I really do not understand why with a fix number of visit a better hardware would play better.
For sure would play faster but not better.

**Comment by StanTraykov** [comment]
@marcocalignano Of course, I agree, it would not play better with fixed visits and I don't think I have ever said anything to that effect. It would play the same. Maybe the misunderstanding comes from the times I/we have used loose words like "better" without further qualifying. I'm a go player, not a NN programmer, and when I use "better" in general context, I mean a program that would more likely win against another program/player. My post above was in reply to the original issue and more or less in agreement with your argument that a large network shouldn't necessarily be punished because of it.

**Comment by Mardak** [comment]
@marcocalignano The switch to 128x10 was at the same time the server switched from 1600 playouts to 3200 visits. https://github.com/gcp/leela-zero/issues/965#issuecomment-370382016

The new 3200 visits was chosen because @killerducky measured that 1600 playouts is pretty close to 5200 visits for 128x6, so accounting for the 10/6 slower 128x10 networks results in 3200 visits. https://github.com/gcp/leela-zero/issues/965#issuecomment-370500844

Just because matches haven't allowed mis-matched visits/playouts in the past doesn't mean we wouldn't want it going forwards. It's just that the number of times we have needed to match up differently sized networks is much fewer than the matches that have happened so far.

So I suppose on that note, if this issue is truly undesired to be part of autogtp as even if it's valuable, it'll be infrequently used, which potentially makes the code harder to maintain, then we'll probably just estimate the ELO difference and adjust the visits as a one-off measurement whenever the time comes to switch sizes.

**Comment by Dorus** [comment]
Idk but this seems like a good idea.

It seems really arbitrary to say "Network A is 5x slower so it needs +200 elo to compensate" instead of "Network A is 5x slower so it needs 50% winrate on 5x less visits".

Both tests are not optimal, because in the first case you still need to know how well the networks scale with more or less playouts/time, and generally going from 100 to 500 visits gives more elo boost than 1000 to 5000. In the second case it's also arbitrary because you need to decide how many playouts to use, do you go 640 vs 3200 or 3200 vs 16000?

Anyway, the second test with different playouts seems more accurate. If anything, having the ability to do a test like that would even allow us to test how much playouts affect elo!

And beside, adding this option to autogtp doesn't even mean we have to use it, that is a decision that gcp can take at a later date. For some matches it might make sense to use unbalanced playouts, for some it might not make sense.

**Comment by dtracers** [comment]
I really think that the better solution would be to just limit the networks on time and do dynamic visits per turn.  Having different visits is really just a way to simulate time.

So why don't we have the test severs perform a visits to time calculation for the current network.
Then limit all future test matches to that time limit for both networks?

The real reason for time limits is strength in a real world setting against a human and making that curve maximum.

**Comment by Mardak** [comment]
> Having different visits is really just a way to simulate time.

Yes, visits and playouts are approximating time in a way that is comparable across different machinesgenerating games and matches. Using actual time run into problems of different hardware capabilities and dependent on what else is happening (e.g., another instance also using the GPU).

As @StanTraykov most recently highlighted in https://github.com/gcp/leela-zero/issues/1113#issuecomment-377679752, different hardware capabilities have very different results when using time as well as different network sizes have different ratios of speedup/slowdown depending on the hardware.

**Comment by dtracers** [comment]
yes this is why I was saying that when we start up the game we find that particular machines visits to time ratio for the current network.

**Comment by Mardak** [comment]
In thinking about #1177, I realized that asymmetrical promotion matches (more visits for one network) could be used for same-sized networks to bias win rates without changing the 55% / 400 game threshold.

For example, if the challenging network had fewer visits, it would need to be more efficient in its search to win games. Similarly, if the challenging network had half the visits, it would effectively need to overcome a 200 ELO handicap.

On the flip side, one way to nudge towards always promote is to give the challenger network additional visits. This would inflate the ELO graph by making an actually-same-strength network get a higher dot on the graph. Doing this would help fresh networks get promoted and on the surface appearing to have upwards progression.

Using the twice-as-fast = 200 ELO metric, giving 3600 visits vs 3200 baseline would roughly give 34 ELO / 55% win rate. Although at that point, why not just lower the win rate threshold from 55% instead of muddying the ELO data…

Looks like we've moved from 128x10 to 192x15 without much need for this issue as we just estimate with the expected slowdown to ELO calculation. So perhaps this is WONTFIX.
